sha256-avx2/utilsx8.c

```diff
@@ -5,121 +5,142 @@
-                    stackx8 + 2*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 3*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 4*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 5*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 6*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 7*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 0*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 1*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 2*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 3*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 4*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 5*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 6*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N,
-                    stackx8 + 7*(tree_height + 1)*SPX_N + (offset - 2)*SPX_N, 2, pub_seed, tree_addrx8);
-            offset--;
-            /* Note that the top-most node is now one layer higher. */
-            heights[offset - 1]++;
 
-            /* If this is a node we need for the auth path.. */
+            /* Ok, we're at a right node (or doing the top 3 levels) */
+            /* Now combine the left and right logical nodes together */
+
+            /* Set the address of the node we're creating. */
+            int j;
+            internal_idx_offset >>= 1;
             for (j = 0; j < 8; j++) {
-                if (((leaf_idx[j] >> heights[offset - 1]) ^ 0x1) == tree_idx) {
-                    memcpy(auth_pathx8 + j*tree_height*SPX_N + heights[offset - 1]*SPX_N,
-                           stackx8 + j*(tree_height + 1)*SPX_N + (offset - 1)*SPX_N, SPX_N);
-                }
+                set_tree_height(tree_addrx8 + j*8, h + 1);
+                set_tree_index(tree_addrx8 + j*8,
+                     (8/2) * (internal_idx&~1) + j - left_adj + internal_idx_offset );
             }
+            unsigned char *left = &stackx8[h * 8 * SPX_N];
+            thashx8( &current[0 * SPX_N],
+                     &current[1 * SPX_N],
+                     &current[2 * SPX_N],
+                     &current[3 * SPX_N],
+                     &current[4 * SPX_N],
+                     &current[5 * SPX_N],
+                     &current[6 * SPX_N],
+                     &current[7 * SPX_N],
+                     &left   [0 * SPX_N],
+                     &left   [2 * SPX_N],
+                     &left   [4 * SPX_N],
+                     &left   [6 * SPX_N],
+                     &current[0 * SPX_N],
+                     &current[2 * SPX_N],
+                     &current[4 * SPX_N],
+                     &current[6 * SPX_N],
+                     2, pub_seed, tree_addrx8);
```

Comment:
sha256x8 is slower than a single sha256, but much faster than 8 times a normal sha256.  What is the crossover point?  If it's at 4 then it might be worthwhile to do these last three layers separately with sha256 (instead of the x8 variant.)  If we remove that from the main loop of this function, then it might be easier to understand as well (as we can get rid of left_adj.)  The additional code might again make it slower though.  The differences either way won't be huge although I wouldn't mind any improvements in readability here.